thinc/loss.py

```diff
@@ -58,142 +171,190 @@ def __init__(
+        xp = get_array_module(guesses)
+        # FIXME: convert using ops?
+        xp_truths = cast(Ints1d, xp.asarray(truths, dtype="i"))
+        truths_2d = to_categorical(
+            xp_truths, n_classes=guesses.shape[-1], label_smoothing=self.label_smoothing
+        )
+        mask = _make_mask(guesses, missing)
+        return cast(Floats2d, truths_2d), mask
+
+    def _convert_strs(
+        self, guesses: Floats2d, truths: Sequence[str]
+    ) -> Tuple[Floats2d, Floats2d]:
+        """
+        Convert Sequence[int] into a Floats2d one-hot array.
+        """
+
         missing_value = self.missing_value
-        # Convert list of ints or list of strings
-        if isinstance(truths, list):
-            truths = list(truths)
-            if len(truths):
-                if isinstance(truths[0], int):
-                    for i, value in enumerate(truths):
-                        if value == missing_value:
-                            missing.append(i)
-                else:
-                    if self.names is None:
-                        msg = (
-                            "Cannot calculate loss from list of strings without names. "
-                            "You can pass the names as a keyword argument when you "
-                            "create the loss object, "
-                            "e.g. CategoricalCrossentropy(names=['dog', 'cat'])"
-                        )
-                        raise ValueError(msg)
-                    for i, value in enumerate(truths):
-                        if value == missing_value:
-                            truths[i] = self.names[0]
-                            missing.append(i)
-                        elif (
-                            value
-                            and self.neg_prefix
-                            and value.startswith(self.neg_prefix)
-                        ):
-                            truths[i] = value[len(self.neg_prefix) :]
-                            neg_index = self._name_to_i[truths[i]]
-                            negatives_mask[i] = 0  # type: ignore
-                            negatives_mask[i][neg_index] = -1  # type: ignore
-                    truths = [self._name_to_i[name] for name in truths]
-            truths = xp.asarray(truths, dtype="i")
-            mask = _make_mask(guesses, missing)
-        else:
-            mask = _make_mask_by_value(truths, guesses, missing_value)
-        if truths.ndim != guesses.ndim:
-            # transform categorical values to one-hot encoding
-            truths = to_categorical(
-                cast(Ints1d, truths),
-                n_classes=guesses.shape[-1],
+        if self.names is None:
+            raise ValueError(
+                "Cannot calculate loss from Sequence[str] without names. "
+                "You can pass the names as a keyword argument when you "
+                "create the loss object, "
+                "e.g. CategoricalCrossentropy(names=['dog', 'cat'])"
```

Comment:
Seems a bit too verbose to me to include an example. Think we can trim it down to the following?

```suggestion
                "create the loss object"
```